### DISCUSSION TITLE
Help with json_each

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#2349] [Q&A]
I'm trying to get two `points`  named as something different in the same record.  Each `site` has a 2 multi-select.  This is sort of working except that it's listing two rows per site.  Somehow I need to get each point renamed like pInt, pExt for each row.

```sql
SELECT
  us.site as id,
  p.name,
  p.type as pType,
  s.created,
  u.id as user,
  s.status,
  s.type as sType
FROM users_sites us
LEFT JOIN sites s on us.site = s.id
LEFT JOIN users u on us.user = u.id
INNER JOIN json_each(s.points) as je_p
INNER JOIN points p ON p.id = je_p.value 
```

To clarify, I'm already able to do something that is working using API Filters of normal collections.  I'm trying to see how far that I can stretch the custom SQL views.

**Comment by dschissler** [comment]
I was trying to do a subquery somehow:

```sql
SELECT
  us.site as id,
  pInt.name as nameInt,
  pExt.name as nameExt,
  s.created,
  u.id as user,
  s.status,
  s.type as sType
FROM users_sites us
LEFT JOIN sites s on us.site = s.id
LEFT JOIN users u on us.user = u.id
LEFT JOIN
  (
  SELECT p.name
  FROM json_each(s.points) as je_p
  INNER JOIN points p ON p.id = je_p.value 
  WHERE je_p.value = 'int'
) as pInt ON 
	
```

**Reply by ganigeorgiev** [reply]
The question is too vague and I'm not sure that I understand what is the expected outcome.

Could you elaborate a little more on your collections structure and provide an example data with the desired format?

**Reply by dschissler** [reply]
Sites
   -> Points (Sites has 2-select points), has name field

I want output like

`site_id point1_name point2_name`

**Reply by ganigeorgiev** [reply]
If you have a `sites` collection and you know that each "site" has 2 `points` related records, then you can use [`json_extract`](https://www.sqlite.org/json1.html#jex) to target each index. For example:

```sql
SELECT DISTINCT
  us.site as id,
  p1.name as p1Name,
  p2.name as p2Name
FROM users_sites us
LEFT JOIN sites s on us.site = s.id
LEFT JOIN points p1 on p1.id = json_extract(s.points, "$[0]")
LEFT JOIN points p2 on p2.id = json_extract(s.points, "$[1]")
```

But there are not enough details to understand what are the relations between the described collections, what they represents, nor what are their available fields.

**Reply by dschissler** [reply]
I wasn't able to get it to accept it with the `DISTINCT`.  When I removed it then I got output but like in another situation it only listed `p2Name` and not `p1Name`.

**Reply by ganigeorgiev** [reply]
> I wasn't able to get it to accept it with the DISTINCT

Hm, this could be a bug with the select columns parser. I'll have a look at it sometime later today.

**Reply by ganigeorgiev** [reply]
@dschissler Yes, the `DISTINCT` is not parsed properly. I'll try to submit a fix a little bit later. As workaround you can try with `GROUP BY id` statement.

**Reply by dschissler** [reply]
@ganigeorgiev Do you mean this?  If so it gives me: `invalid query - empty table info probably due to invalid or missing table _temp_ZQsSP`

```sql
SELECT
  us.site as id,
  p1.name as p1Name,
  p2.name as p2Name
FROM users_sites us
LEFT JOIN sites s on us.site = s.id
LEFT JOIN points p1 on p1.id = json_extract(s.points, "$[0]")
LEFT JOIN points p2 on p2.id = json_extract(s.points, "$[1]")
GROUP BY id
```

**Reply by ganigeorgiev** [reply]
Yes, but note that the above was just an example and I'm not sure if the correct relations or fields are used.

**Reply by dschissler** [reply]
@ganigeorgiev I've been getting that missing table message a lot.  Sometimes it's catching a silly mistake, but other times I really don't know if it's a bug or my mistake (particularly around sub-queries).  My SQL is rusty.

**Reply by dschissler** [reply]
I mixed it around a bit.  It seems that GROUP BY is not allowing me to use a field alias.

This works (gives undesired output but compiles)

```sql
SELECT
  s.id as id,
  p1.name as p1Name,
  p2.name as p2Name
FROM users_sites us
LEFT JOIN sites s on us.site = s.id
LEFT JOIN points p1 on p1.id = json_extract(s.points, "$[0]")
LEFT JOIN points p2 on p2.id = json_extract(s.points, "$[1]")
GROUP BY s.id
```

but this does not

```sql
SELECT
  s.id as id,
  p1.name as p1Name,
  p2.name as p2Name
FROM users_sites us
LEFT JOIN sites s on us.site = s.id
LEFT JOIN points p1 on p1.id = json_extract(s.points, "$[0]")
LEFT JOIN points p2 on p2.id = json_extract(s.points, "$[1]")
GROUP BY id
```

**Reply by ganigeorgiev** [reply]
@dschissler Yes it is possible that the `id` column is treated as ambiguous and require a more detailed identifier.

_Just to let you know that I've pushed in [`master`](https://github.com/pocketbase/pocketbase/commit/cae3315e463b362b05ac0f369cdbcab7af97052d) a fix for the `DISTINCT` parsing. Since the `GROUP BY` could be used as an alternative I will not rush a release, but eventually it will be available with the next minor v0.15.2._

**Reply by dschissler** [reply]
@ganigeorgiev I checked these queries via command line `sqlite3`.  Both are giving all three fields.  My concern is that there will still be the bug where it's not outputing `p1Name` but only `p2Name` (when I used `GROUP BY s.id`).

**Reply by ganigeorgiev** [reply]
I'm not sure how to reproduce it. Please provide a minimal reproducible repo with sample data.

Internally the view query is a plain SQL view, so if you were able to save the collection it should behave the same.

**Reply by dschissler** [reply]
@ganigeorgiev What is the best way for me to give this to you?  It seems like a single or multi-select collection export feature would be useful.

**Reply by dschissler** [reply]
@ganigeorgiev 

Points: has name
Sites: has 2-relation to Points
User_Sites: has 1-relation to Sites, has 1-relation to user.

**Reply by ganigeorgiev** [reply]
@dschissler Please create a minimal reproducible github repo with the sample `pb_data` committed.

**Reply by dschissler** [reply]
@ganigeorgiev 

Also I think that I'm going to need `json_each` because I can't be sure if the first item is type `int` or `ext`.  It's possible that what I was doing before was correct but that there was just a bug.  I was sure confused.

bob@email.com
openopenopen
https://github.com/dschissler/pocketbase_bug

**Reply by ganigeorgiev** [reply]
Thank you. I understand now what is the issue and why you are getting only the second p2Name.

While the underlying view query is the same as it is saved in the collection, the generated collection schema is not properly resolving all fields. The reason for this is because both `p1Name` and `p2Name` originate from the same field source (aka. `points.name`) and they have the same field id and when generating the schema fields slice the second p2Name field is end up replacing the previously inserted p1Name. I'll try to submit a fix sometime later today and will update the test cases.

**Reply by ganigeorgiev** [reply]
@dschissler I've submitted another fix in [`master`](https://github.com/pocketbase/pocketbase/commit/e61d48fe7b438c42a1f3e5df49c43085f12478b9). If you are using PocketBase as framework and want to test it, you can update your `go.mod` to point to the latest master commit (or run `go get github.com/pocketbase/pocketbase@master`).
I'm planning a final v0.15.2 release sometime later this week.

**Reply by dschissler** [reply]
@ganigeorgiev Two big bug fixes in one little version.  Not bad.

**Reply by dschissler** [reply]
@ganigeorgiev 

This is working for me.  Do you know of a better way to do this?  I'm concerned that the json_extract is brittle (because it means that I have to guarantee the ordering).  Damn though, if I'm always going to have two of them then maybe they should just be two single relations for a much easier time.  At any rate, I'd like to explore this a bit more.

```sql
SELECT DISTINCT
  us.site as id,
  pInt.name as nameInt,
  pExt.name as nameExt,
  s.created,
  u.id as user,
  s.status,
  s.type
FROM users_sites us
LEFT JOIN sites s on us.site = s.id
LEFT JOIN users u on us.user = u.id

INNER JOIN json_each(s.points) as je_pint
INNER JOIN points pInt ON pInt.id = je_pint.value

INNER JOIN json_each(s.points) as je_pext
INNER JOIN points pExt ON pExt.id = je_pext.value

WHERE pExt.type = 'ext' AND pInt.type = 'int'
```
